src/lib/types.ts

```diff
@@ -25,19 +25,30 @@ export interface EncodeBuffer {
 }
 
 export interface BACNetAddress {
+	/** 0 for local, 1 for IP, 2 for MAC, etc. */
 	type?: number
+	/** 0 for local, > 0 for remote, 0xffff for broadcast */
 	net?: number
+	/** IP address or MAC address, etc. */
 	adr?: number[]
-	address?: string
-	forwardedFrom?: string
 }
 
-export interface ReceiverAddress {
+export interface DecodedAddress {
 	address: string
 	forwardedFrom?: string
 }
 
-export type AddressParameter = string | ReceiverAddress
+export interface DecodedNpdu {
+	len: number
+	funct: number
+	destination?: BACNetAddress
+	source?: BACNetAddress
+	hopCount: number
+	networkMsgType: number
+	vendorId: number
+}
+
+export type AddressParameter = string | DecodedAddress
```

Comment:
Insofar as possible, I would try to get rid these kinds of unions and only use the complex / parsed version of any given type (in this case I would drop `string` and keep `DecodedAddress`) tweaking everything else accordingly. I can't provide an in-depth review right now so take my words with a few grains of salt.

Reply:
Agree 👍🏼